### PR TITLE
Add planEndEffectorTwist

### LINKED ISSUE #465: Plan trajectories for end-effector twists
There's an implementation for `MoveEndEffectorTwistVectorField` in the [`se2_policies`](https://github.com/personalrobotics/aikido/tree/se2_policies) branch, but we need to clean it up.

### PR DESCRIPTION
**Pull request by nansongyi**
[Describe this pull request. Link to relevant GitHub issues, if any.]

[Explain how this pull request was tested.]

Resolves #465.

***

**Before creating a pull request**

- [ ] Document new methods and classes
- [x] Format code with `make format`

**Before merging a pull request**

- [x] Set version target by selecting a milestone on the right side
- [ ] Summarize this change in `CHANGELOG.md`
- [ ] Add unit test(s) for this change

### CODE REVIEW

File: src/planner/vectorfield/VectorFieldPlanner.cpp [hunk 1]
```diff
@@ -5,6 +5,7 @@
 #include "aikido/constraint/TestableIntersection.hpp"
 #include "aikido/constraint/dart/JointStateSpaceHelpers.hpp"
 #include "aikido/planner/vectorfield/MoveEndEffectorOffsetVectorField.hpp"
+#include "aikido/planner/vectorfield/MoveEndEffectorTwistVectorField.hpp"
```
Comment (brianhou) [inline src/planner/vectorfield/VectorFieldPlanner.cpp]: This header file is missing from this PR.

File: include/aikido/planner/vectorfield/MoveEndEffectorTwistVectorField.hpp [hunk 2]
```diff
@@ -0,0 +1,92 @@
+#ifndef AIKIDO_PLANNER_VECTORFIELD_MOVEENDEFFECTORTWISTVECTORFIELD_HPP_
+#define AIKIDO_PLANNER_VECTORFIELD_MOVEENDEFFECTORTWISTVECTORFIELD_HPP_
+
+#include <aikido/planner/vectorfield/BodyNodePoseVectorField.hpp>
```
Comment (brianhou) [inline include/aikido/planner/vectorfield/MoveEndEffectorTwistVectorField.hpp]:
```suggestion
#include "aikido/planner/vectorfield/BodyNodePoseVectorField.hpp"
```

File: include/aikido/planner/vectorfield/MoveEndEffectorTwistVectorField.hpp [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+#include <aikido/trajectory/Interpolated.hpp>
```
Comment (brianhou) [inline include/aikido/planner/vectorfield/MoveEndEffectorTwistVectorField.hpp]:
```suggestion
#include "aikido/trajectory/Interpolated.hpp"
```

File: src/planner/vectorfield/MoveEndEffectorTwistVectorField.cpp [hunk 4]
```diff
@@ -0,0 +1,69 @@
+#include <dart/dynamics/BodyNode.hpp>
+#include <dart/math/MathTypes.hpp>
+#include <dart/optimizer/Function.hpp>
+#include <dart/optimizer/Problem.hpp>
+#include <aikido/planner/vectorfield/MoveEndEffectorTwistVectorField.hpp>
```
Comment (brianhou) [inline src/planner/vectorfield/MoveEndEffectorTwistVectorField.cpp]:
```suggestion
#include "aikido/planner/vectorfield/MoveEndEffectorTwistVectorField.hpp"
```

File: src/planner/vectorfield/MoveEndEffectorTwistVectorField.cpp [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+#include <aikido/planner/vectorfield/VectorFieldUtil.hpp>
```
Comment (brianhou) [inline src/planner/vectorfield/MoveEndEffectorTwistVectorField.cpp]:
```suggestion
#include "aikido/planner/vectorfield/VectorFieldUtil.hpp"
```

File: include/aikido/robot/util.hpp [hunk 6]
```diff
@@ -200,6 +200,32 @@ trajectory::TrajectoryPtr planToEndEffectorOffset(
...
+/// Plan with a desired end-effector twist.
+/// \param[in] space StateSpace for the metaskeleton
+/// \param[in] metaSkeleton Metaskeleton to plan with
+/// \param[in] body Bodynode for the end-effector
+/// \param[in] twistSeq Twist for the end-effector
+/// \param[in] durationSeq Time to plan with the desired twist
+/// \param[in] collisionTestable Collision constraint to check. Self-collision
+/// is checked by default.
+/// \param[in] timelimit Timelimit for the plan to be generated
+/// \param[in] positionTolerance Tolerance in position // do I need this?
```
Comment (aditya-vk) [inline include/aikido/robot/util.hpp]: What are the units?

File: include/aikido/robot/util.hpp [hunk 7]
```diff
@@ -200,6 +200,32 @@ trajectory::TrajectoryPtr planToEndEffectorOffset(
...
+/// \param[in] angularTolerance Tolerance in angle // do I need this?
```
Comment (aditya-vk) [inline include/aikido/robot/util.hpp]: What are the units?

File: include/aikido/robot/util.hpp (hunk 6 above)
Comment (aditya-vk) [reply include/aikido/robot/util.hpp]: "do I need this?"

File: include/aikido/robot/util.hpp (hunk 7 above)
Comment (aditya-vk) [reply include/aikido/robot/util.hpp]: "do I need this?"

File: include/aikido/robot/util.hpp [hunk 8]
```diff
@@ -200,6 +200,32 @@ trajectory::TrajectoryPtr planToEndEffectorOffset(
     = VectorFieldPlannerParameters(),
     const CRRTPlannerParameters& crrtParameters = CRRTPlannerParameters());
 
+/// Plan with a desired end-effector twist.
+/// \param[in] space StateSpace for the metaskeleton
+/// \param[in] metaSkeleton Metaskeleton to plan with
+/// \param[in] body Bodynode for the end-effector
+/// \param[in] twistSeq Twist for the end-effector
+/// \param[in] durationSeq Time to plan with the desired twist
```
Comment (aditya-vk) [inline include/aikido/robot/util.hpp]: What is the reasoning behind naming `twistSeq` and `durationSeq`? Why not `twist` and `timeout`?

File: include/aikido/robot/util.hpp [hunk 9]
```diff
@@ -200,6 +200,32 @@ trajectory::TrajectoryPtr planToEndEffectorOffset(
...
+/// \param[in] collisionTestable Collision constraint to check. Self-collision
+/// is checked by default.
+/// \param[in] timelimit Timelimit for the plan to be generated
```
Comment (aditya-vk) [inline include/aikido/robot/util.hpp]: How is this different from the docstring for `durationSeq`? Both seem to suggest planning timeout

File: include/aikido/robot/util.hpp [hunk 10]
```diff
@@ -200,6 +200,32 @@ trajectory::TrajectoryPtr planToEndEffectorOffset(
...
+/// \param[in] vfParameters VectorFieldPlanenr parameters
+/// \return Output trajectory
+trajectory::TrajectoryPtr planWithEndEffectorTwist(
+    const statespace::dart::MetaSkeletonStateSpacePtr& space,
+    const dart::dynamics::MetaSkeletonPtr& metaSkeleton,
+    const dart::dynamics::BodyNodePtr& body,
+    const Eigen::Vector6d& twistSeq,
+    double durationSeq,
+    const constraint::TestablePtr& collisionTestable,
+    double timelimit,
```
Comment (aditya-vk) [inline include/aikido/robot/util.hpp]: Take this as `const`

File: include/aikido/robot/util.hpp [hunk 11]
```diff
@@ -200,6 +200,32 @@ trajectory::TrajectoryPtr planToEndEffectorOffset(
...
+    double positionTolerance = 1e-3,
+    double angularTolerance = 1e-3,
```
Comment (aditya-vk) [inline include/aikido/robot/util.hpp]: These should be `const` as well

File: include/aikido/planner/vectorfield/MoveEndEffectorTwistVectorField.hpp [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+
+public:
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+};
+
+} // namespace vectorfield
+} // namespace planner
+} // namespace aikido
+
+#endif // ifndef
```
Comment (aditya-vk) [inline include/aikido/planner/vectorfield/MoveEndEffectorTwistVectorField.hpp]: Weirdly formatted.

File: include/aikido/planner/vectorfield/MoveEndEffectorTwistVectorField.hpp [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+#define AIKIDO_PLANNER_VECTORFIELD_MOVEENDEFFECTORTWISTVECTORFIELD_HPP_
+
+#include "aikido/planner/vectorfield/BodyNodePoseVectorField.hpp"
+#include "aikido/trajectory/Interpolated.hpp"
+
+namespace aikido {
+namespace planner {
+namespace vectorfield {
+
+/// Vector field for moving end-effector by a direction and distance.
```
Comment (aditya-vk) [inline include/aikido/planner/vectorfield/MoveEndEffectorTwistVectorField.hpp]: This docstring needs to be updated for twist.

File: include/aikido/planner/vectorfield/MoveEndEffectorTwistVectorField.hpp [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+///
+/// This class defines two callback functions for vectorfield planner.
+/// One for generating joint velocity in MetaSkeleton state space,
+/// and one for determining vectorfield planner status.
+class MoveEndEffectorTwistVectorField : public BodyNodePoseVectorField
+{
+public:
+  /// Constructor
+  ///
+  /// \param[in] stateSpace MetaSkeleton state space.
```
Comment (aditya-vk) [inline include/aikido/planner/vectorfield/MoveEndEffectorTwistVectorField.hpp]: Update constructor docstring

File: include/aikido/planner/vectorfield/MoveEndEffectorTwistVectorField.hpp [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+  /// limit, velocity is bounded in that direction to 0.
+  MoveEndEffectorTwistVectorField(
+      aikido::statespace::dart::ConstMetaSkeletonStateSpacePtr stateSpace,
+      ::dart::dynamics::MetaSkeletonPtr metaskeleton,
+      ::dart::dynamics::ConstBodyNodePtr bn,
+      const Eigen::Vector6d& twistSeq,
+      double durationSeq,
+      double positionTolerance,
+      double angularTolerance,
+      double maxStepSize,
```
Comment (aditya-vk) [inline include/aikido/planner/vectorfield/MoveEndEffectorTwistVectorField.hpp]: All of these parameters need to be taken as const

File: include/aikido/planner/vectorfield/MoveEndEffectorTwistVectorField.hpp [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+      const Eigen::Isometry3d& pose) const override;
+
+protected:
+  /// Count
+  mutable int mCount{0};
+
+  /// Twist specified
+  Eigen::Vector6d mTwist;
+
+  /// Duration specified
```
Comment (aditya-vk) [inline include/aikido/planner/vectorfield/MoveEndEffectorTwistVectorField.hpp]: What does this duration imply? I think a more detailed docstring would be beneficial.

File: include/aikido/planner/vectorfield/MoveEndEffectorTwistVectorField.hpp [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+  double mDuration;
+
+  /// Do we need this at this point?
```
Comment (aditya-vk) [inline include/aikido/planner/vectorfield/MoveEndEffectorTwistVectorField.hpp]: Update docstring.

File: include/aikido/planner/vectorfield/MoveEndEffectorTwistVectorField.hpp [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+  aikido::trajectory::InterpolatedPtr mTwistTraj;
+
+  /// Tolerance of linear deviation error.
+  double mPositionTolerance;
+
+  /// Tolerance of angular error.
```
Comment (aditya-vk) [inline include/aikido/planner/vectorfield/MoveEndEffectorTwistVectorField.hpp]: What are the units?

File: include/aikido/planner/vectorfield/MoveEndEffectorTwistVectorField.hpp [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+  /// Twist specified
+  Eigen::Vector6d mTwist;
+
+  /// Duration specified
+  double mDuration;
+
+  /// Do we need this at this point?
+  aikido::trajectory::InterpolatedPtr mTwistTraj;
+
+  /// Tolerance of linear deviation error.
```
Comment (aditya-vk) [inline include/aikido/planner/vectorfield/MoveEndEffectorTwistVectorField.hpp]: What are the units?

File: include/aikido/planner/vectorfield/MoveEndEffectorTwistVectorField.hpp [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+  /// \param[in] maxStepSize The maximum step size used to guarantee
+  /// that the integrator does not step out of joint limits.
+  /// \param[in] jointLimitPadding If less then this distance to joint
+  /// limit, velocity is bounded in that direction to 0.
+  MoveEndEffectorTwistVectorField(
+      aikido::statespace::dart::ConstMetaSkeletonStateSpacePtr stateSpace,
+      ::dart::dynamics::MetaSkeletonPtr metaskeleton,
+      ::dart::dynamics::ConstBodyNodePtr bn,
+      const Eigen::Vector6d& twistSeq,
+      double durationSeq,
```
Comment (aditya-vk) [inline include/aikido/planner/vectorfield/MoveEndEffectorTwistVectorField.hpp]: This variable can be renamed better

File: include/aikido/planner/vectorfield/MoveEndEffectorTwistVectorField.hpp [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+  /// \param[in] angularTolerance Constraint tolerance in radians.
+  /// \param[in] maxStepSize The maximum step size used to guarantee
+  /// that the integrator does not step out of joint limits.
+  /// \param[in] jointLimitPadding If less then this distance to joint
+  /// limit, velocity is bounded in that direction to 0.
+  MoveEndEffectorTwistVectorField(
+      aikido::statespace::dart::ConstMetaSkeletonStateSpacePtr stateSpace,
+      ::dart::dynamics::MetaSkeletonPtr metaskeleton,
+      ::dart::dynamics::ConstBodyNodePtr bn,
+      const Eigen::Vector6d& twistSeq,
```
Comment (aditya-vk) [inline include/aikido/planner/vectorfield/MoveEndEffectorTwistVectorField.hpp]: This variable can be renamed better

File: src/planner/vectorfield/MoveEndEffectorTwistVectorField.cpp [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+
+//==============================================================================
+VectorFieldPlannerStatus
+MoveEndEffectorTwistVectorField::evaluateCartesianStatus(
+    const Eigen::Isometry3d& pose) const
+{
+  mCount += 1;
+  DART_UNUSED(pose);
+
+  // (avk): Not computing the error/deviation for now
```
Comment (aditya-vk) [inline src/planner/vectorfield/MoveEndEffectorTwistVectorField.cpp]: Hmm? Can we address these TODOs? They are important.

**Comment by brianhou** [comment]
@nansongyi Why did we close this PR? My impression was that it was fairly close to being done. I'm ok with closing this if it's just in preparation for reapplying those changes on top of the formatting changes in the current master branch or something, but I don't want to waste your efforts!

**Comment by nansongyi** [comment]
oh I am sorry. I think I misunderstood something. 
When Ethan closed this pr, [libada#32](https://github.com/personalrobotics/libada/pull/32#issuecomment-683920194), I thought it's the one for aikido, so I ended up with ignoring it cuz I thought it's already closed.
Today, when I clean up my workspace, I deleted this branch `squirrel_pr ` cuz I thought it's already closed. 
I will talk to Ethan whether we still need it or not, then I will get back to you.
Thanks.

**Comment by brianhou** [comment]
Reopening for now, just to make sure we don't forget/lose this 🙂

**Comment by egordon** [comment]
WONTFIX for now. From an application standpoint, planToEndEffector<Offset/Twist> only adds collision checking compared to the new JacobianVelocityController (adding in #605).

We can re-open if requested.

**Pull request closed without merging**